### DISCUSSION TITLE
Does ```convert_sh_to_legacy``` work as intended ?

### DISCUSSION THREAD

**Discussion by DelinteNicolas** [dipy/dipy#2677] [Reconstruction]
## Description

I was converting ODFs files, saved using spherical harmonics (SH), from the MRtrix3 format (Tournier) to DIPY (Descoteaux) and had issue with this function. It currently flips the SH coefficients when m is odd and <0 : ```out_sh_coeffs = sh_coeffs * np.where(m < 0, (-1.)**m, 1.)```.
However, I noticed that the coeffifient that I needed to flip to get a working files were the 'even' m numbers. As mentionned in the DIPY theory (https://dipy.org/documentation/1.4.0./theory/sh_basis/) : "the effect of this change is a sign flip for the SH functions of even degree m<0". 

If I understand correctly, shouldn't the code line be replaced with  ```out_sh_coeffs = sh_coeffs * np.where(m < 0, -1*(-1.)**m, 1.)``` ?

With:
DIPY==1.5.0
MRtrix==3.0.2

**Comment by skoudoro** [comment]
Yes, it worked as intended. 

In your case, you might want to use `convert_sh_to_full_basis` instead.

@CHrlS98 can tell you more when he has time.

**Comment by CHrlS98** [comment]
Hi @DelinteNicolas,
If your need is simply converting your SH coefficients from the `Tournier` basis to the `Descoteaux` basis, you should not need `convert_sh_to_legacy`. This function should be used, for instance, to convert MRtrix3 SH coefficients to earlier MRtrix0.2 SH coefficients. Also, `convert_sh_to_full_basis` is instead used to convert coefficients expressed in a symmetric SH basis to a full SH basis comprising even and odd order SH coefficients.

As such, none of those methods converts between `Tournier` and `Descoteaux` bases. As far as I know there is no such function in DIPY as of now (maybe @skoudoro would know better), but I agree it could be useful.

Let me know if I misunderstood your problem or if you need further assistance!

**Comment by jhlegarreta** [comment]
@DelinteNicolas The `test_real_sh_descoteaux1` method seems to test a case where the  legacy Tournier basis is converted to the legacy Descoteaux basis:
https://github.com/dipy/dipy/blob/12f3324715b970df4999cca27deb85886ae511e4/dipy/reconst/tests/test_shm.py#L187

And it does not seem that your reasoning here:
> out_sh_coeffs = sh_coeffs * np.where(m < 0, -1*(-1.)**m, 1.)

would be successful: if we compute the new bases in the named test for the first 15 coefficients (the first theta, phi value):

```
new_basis_tournier = convert_sh_from_legacy(basis[0], "tournier07")
expected *= np.where(m == 0, 1., np.sqrt(2))
new_basis_descoteaux = convert_sh_from_legacy(expected[0], "descoteaux07")
```

The following is false:
```
npt.assert_equal (new_basis_descoteaux, new_basis_tournier * np.where(m < 0, -1 * (-1.) ** m, 1.))
```

So unless I am missing something, the conversion function from `tournier07` to `descoteaux07` (according to the definitions in DIPY>1.3) is not `np.where(m < 0, -1 * (-1.) ** m, 1.)`: the parts of the SH taken (real or imaginary) for each range of the degree `m` change across both bases (both in the legacy and current definitions).

**Reply by jhlegarreta** [reply]
Now, there might be something else going on in all this: adding in the named test
``` 
descoteaux07_basis_new, _, _ = real_sh_descoteaux(4, sphere.theta, sphere.phi, full_basis=False, legacy=False)
descoteaux07_basis_from_legacy = convert_sh_from_legacy(expected, "descoteaux07", full_basis=False)
assert_array_almost_equal(descoteaux07_basis_new, descoteaux07_basis_from_legacy)

tournier07_basis_new, _, _ = real_sh_tournier(4, sphere.theta, sphere.phi, full_basis=False, legacy=False)
tournier07_basis_from_legacy = convert_sh_from_legacy(basis, "tournier07", full_basis=False)
assert_array_almost_equal(tournier07_basis_new, tournier07_basis_from_legacy)
```

passes for "descoteaux07" but fails for "tournier07". Not sure if I am missing something.
